helios-services/src/main/java/com/spotify/helios/master/ZooKeeperMasterModel.java

```diff
@@ -893,42 +938,176 @@ private RollingUpdateOp rollingUpdateDeploy(final ZooKeeperClient client,
-          deployment.getDeploymentGroupName(), deploymentGroup.getName());
-      final boolean isSameJob = deployment.getJobId().equals(deploymentGroup.getJobId());
-      final RolloutOptions rolloutOptions = deploymentGroup.getRolloutOptions();
-
-      if (isOwnedByDeploymentGroup || (
-          isSameJob && rolloutOptions.getMigrate())) {
-        if (isSameJob && isOwnedByDeploymentGroup && deployment.getGoal().equals(Goal.START)) {
-          // The job we want deployed is already deployed and set to run, so just leave it.
-          continue;
-        }
 
-        try {
-          final String token = MoreObjects.firstNonNull(
-              deploymentGroup.getRolloutOptions().getToken(), Job.EMPTY_TOKEN);
-          operations.addAll(getUndeployOperations(client, host, deployment.getJobId(), token));
-        } catch (TokenVerificationException e) {
-          return opFactory.error(e, host, RollingUpdateError.TOKEN_VERIFICATION_ERROR);
-        } catch (HostNotFoundException e) {
-          return opFactory.error(e, host, RollingUpdateError.HOST_NOT_FOUND);
-        } catch (JobNotDeployedException e) {
-          // probably somebody beat us to the punch of undeploying. that's fine.
-        }
+      if (!ownedByDeploymentGroup(deployment, deploymentGroup)) {
+        continue;
+      }
+
+      if (!force && redundantDeployment(deployment, deploymentGroup)) {
+        continue;
+      }
+
+      try {
+        final String token = MoreObjects.firstNonNull(
+            deploymentGroup.getRolloutOptions().getToken(), Job.EMPTY_TOKEN);
+        operations.addAll(getUndeployOperations(client, host, deployment.getJobId(), token));
+        log.debug("planned undeploy operations for job={}", deployment.getJobId());
+      } catch (TokenVerificationException e) {
+        return opFactory.error(e, host, RollingUpdateError.TOKEN_VERIFICATION_ERROR);
+      } catch (HostNotFoundException e) {
+        return opFactory.error(e, host, RollingUpdateError.HOST_NOT_FOUND);
+      } catch (JobNotDeployedException e) {
+        // probably somebody beat us to the punch of undeploying. that's fine.
       }
     }
 
     return opFactory.nextTask(operations);
   }
 
+  private boolean ownedByDeploymentGroup(final Deployment deployment,
+                                         final DeploymentGroup deploymentGroup) {
+    // This deployment was created by this deployment group.
+    if (Objects.equals(deployment.getDeploymentGroupName(), deploymentGroup.getName())) {
+      return true;
+    }
+
+    // This deployment was not created by this deployment group, but should be migrated into it
+    // because a migration was requested and it is of the deployment group's job.
+    if (deploymentGroup.getRolloutOptions().getMigrate() &&
+        deployment.getJobId().equals(deploymentGroup.getJobId())) {
+      return true;
+    }
+
+    return false;
+  }
+
+  private boolean redundantDeployment(final Deployment deployment,
```

Comment:
Having renamed this method to `redundantUndeployment` and added a docstring I strongly feel my current implementation is clearer than this suggestion. I think it makes it more obvious *why* we're doing these checks.

As someone new-ish to the codebase I had a really hard time parsing the existing code and what its purpose was. It would be obvious to me from your proposed docstring *what* the desired state was, but not *why*.